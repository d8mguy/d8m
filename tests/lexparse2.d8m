
// test a second grammar incl some testing of the scan methods

load("modules/lexparse")      // Get all the symbols, not just the exported ones

// concat these with wgpure in the absence of d8m grmr
val dependsProdns: list(rawProdn) = [
    ["identlist", :general, "ident"],
    ["identlist", :general, "ident , identlist*"],
    ["exprlist", :general, "expr"],
    ["exprlist", :general, "expr , exprlist*"],
    ["funcall", :expr, "ident ( )"],
    ["funcall", :expr, "ident ( exprlist )"],
    ["stmtlist", :general, "stmt"],
    ["stmtlist", :general, "stmt ;"],
    ["stmtlist", :general, "stmt ; stmtlist*"],
    ["stmts", :stmt, "{ }"],
    ["stmts", :stmt, "{ stmtlist* }"]
]

val wgpureProdns: list(rawProdn) = [
    ["listx", :expr, "list ( ident )"],
    ["listx", :expr, "list ( listx )"],
    ["typex", :general, "ident"],
    ["typex", :general, "listx"],
    ["colonpair", :general, "ident : typex"],
    ["colonpairlist", :general, "colonpair"],
    ["colonpairlist", :general, "colonpair , colonpairlist*"],
    ["identopttyp", :general, "ident"],
    ["identopttyp", :general, "colonpair"],
    ["identopttyplist", :general, "identopttyp"],
    ["identopttyplist", :general, "identopttyp , identopttyplist*"],
    ["responselit", :general, "{ identlist }"],
    ["getpost", :general, "GET"],
    ["getpost", :general, "POST"],
    ["request", :general, "getpost ( )"],
    ["request", :general, "getpost ( identopttyplist )"],
    ["respx", :general, "ident"],
    ["respx", :general, "responselit"],
    ["respx", :general, "( ident , ident )"],
    ["respx", :general, "( ident , responselit )"],
    ["responsex", :general, "response respx"],
    ["responsex", :general, "response respx or respx"],
    ["responsex", :general, "response respx or respx or respx"],
    ["transitionstmt", :stmt, "transition ident strg => request then ident"],
    ["transitionstmt", :stmt, "transition ident strg => request then ident responsex"],
    ["statestmt", :stmt, "state colonpairlist"],
    ["psx", :general, "funcall"],
    ["psx", :general, "ident stmts"],
    ["psx", :general, "funcall stmts"],
    ["pgstmt", :stmt, "pagespec psx"],
    // Both javascript and css are not parseable with these definitions since stmts doesn't include javascript or css
    // constructs, but they are scannable since {} is a bracket pair.
    ["jsstmt", :stmt, "javascript stmts"],
    ["cssstmt", :stmt, "css stmts"],
    ["respstmt", :stmt, "response ident responselit"],
    ["compstmt", :stmt, "components strg"],
    ["initstmt", :stmt, "initial ident , psx"]
]

val wgDefn = [langDefn: [], dependsProdns + wgpureProdns, [],
    ["transition", "state", "pagespec", "javascript", "css", "then", "response", "components", "initial", "GET", "POST", "list", "or"],
    d8mBrackets, d8mComments, d8mSeparators, true]

// long string...
val testspec = `

transition SaveFile "save" => POST(newFilename: string, deletedCols, fieldnames, fielddescs, fieldunits)
    then saveFile response (editPage, fileResponseObj)

// this is for an incrl xhr called when a missingCode is edited on the cleanup page
transition cleanupMissingCode1 "editMCPre" => GET(index: integer, value: string) then editMissingPre
    response {anomalous}

// open a file and either go straight to edit or first to cleanup
transition FileResponse "open" => GET(filepath: string, filename: string) then openFile
    response (editPage, fileResponseObj) or (cleanupPage, cleanupResponse)

// this very similar transition is for when a missingCode is edited on the edit page
transition cleanupMissingCode2 "editMC" => GET(index: integer, value: string) then editMissing
    response {fieldfacts}

javascript {
    import { Delete } from 'react-feather'
}

state deletedCols: list(integer), nCols: integer
state anychange: boolean, fileResponseIndex: integer
state factoids: list(string), fieldnames: list(string), fielddescs: list(string)
state fieldqtys: list(string), fieldbaseunits: list(string), fieldunits: list(string), fieldfacts: list(list(string))
state filename: string, fieldcats: list(string)
state dataTable: list(list(string))
state missingCodes: list(string), anomalous: list(integer), excluded: list(integer)

response fileResponseObj {fileResponseIndex, factoids, missingCodes, nCols, dataTable,
                 fieldnames, fielddescs, fieldunits, fieldcats, fieldfacts}
response cleanupResponse {fileResponseIndex, factoids, missingCodes, nCols, dataTable,
                 anomalous, excludedIndices}

// user excludes a line on cleanup page
transition cleanupExcluded "editExcl" => GET(linenum: integer) then editExcluded

// user says cleanup is done
transition finishCleanup "cleanedup" => GET() then cleanupDone response (editPage, fileResponseObj)

transition CloseFile "close" => GET() then closeFile response startPage

`
// ...end of long string

val driver = \imp() {
    var ps = [parseState: wgDefn]
    ps.bytes = testspec; ps.cxt.clear()
    ps.nxToken()
    while(ps.cxt.toktype != "end") {
        val tkn = ps.cxt.tknval
        val b4 = ps.cxt.ofs - tkn.count
        if(tkn == "css" || tkn == "javascript") {
            ps.scanStmt()
            val after = ps.prevtkn.ofs
            val strgform = b4 <= after ? cvt(ps.bytes[b4...after], string) : "!!!!!"
            println("start: #{b4}; end: #{after}", strgform)
        } else {
            var trm = ps.parseStmt()
            val after = ps.prevtkn.ofs
            var exit = false
            if(trm == nil) { trm = ps.error; exit = true }
            println("start: #{b4}; end: #{after}", trm == nil ? "nil" : stringify(trm))
            if(exit) break
        }
    }
}
driver()
