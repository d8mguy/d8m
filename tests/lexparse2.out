start: 2; end: 161 [transitionstmt: [ident: SaveFile];[strg: "save"];[request: [getpost: [ident: POST]];[identopttyplist: [identopttyp: [ident: newFilename];[typex: [ident: string]]];[identopttyplist: [identopttyp: [ident: deletedCols]];[identopttyp: [ident: fieldnames]];[identopttyp: [ident: fielddescs]];[identopttyp: [ident: fieldunits]]]]];[ident: saveFile];[responsex: [respx: [ident: editPage];[ident: fileResponseObj]]]]
start: 247; end: 372 [transitionstmt: [ident: cleanupMissingCode1];[strg: "editMCPre"];[request: [getpost: [ident: GET]];[identopttyplist: [identopttyp: [ident: index];[typex: [ident: integer]]];[identopttyplist: [identopttyp: [ident: value];[typex: [ident: string]]]]]];[ident: editMissingPre];[responsex: [respx: [responselit: [identlist: [ident: anomalous]]]]]]
start: 440; end: 602 [transitionstmt: [ident: FileResponse];[strg: "open"];[request: [getpost: [ident: GET]];[identopttyplist: [identopttyp: [ident: filepath];[typex: [ident: string]]];[identopttyplist: [identopttyp: [ident: filename];[typex: [ident: string]]]]]];[ident: openFile];[responsex: [respx: [ident: editPage];[ident: fileResponseObj]];[respx: [ident: cleanupPage];[ident: cleanupResponse]]]]
start: 689; end: 809 [transitionstmt: [ident: cleanupMissingCode2];[strg: "editMC"];[request: [getpost: [ident: GET]];[identopttyplist: [identopttyp: [ident: index];[typex: [ident: integer]]];[identopttyplist: [identopttyp: [ident: value];[typex: [ident: string]]]]]];[ident: editMissing];[responsex: [respx: [responselit: [identlist: [ident: fieldfacts]]]]]]
start: 811; end: 868 javascript {
    import { Delete } from 'react-feather'
}
start: 870; end: 918 [statestmt: [colonpairlist: [colonpair: [ident: deletedCols];[typex: [listx: [ident: integer]]]];[colonpairlist: [colonpair: [ident: nCols];[typex: [ident: integer]]]]]]
start: 919; end: 971 [statestmt: [colonpairlist: [colonpair: [ident: anychange];[typex: [ident: boolean]]];[colonpairlist: [colonpair: [ident: fileResponseIndex];[typex: [ident: integer]]]]]]
start: 972; end: 1052 [statestmt: [colonpairlist: [colonpair: [ident: factoids];[typex: [listx: [ident: string]]]];[colonpairlist: [colonpair: [ident: fieldnames];[typex: [listx: [ident: string]]]];[colonpair: [ident: fielddescs];[typex: [listx: [ident: string]]]]]]]
start: 1053; end: 1170 [statestmt: [colonpairlist: [colonpair: [ident: fieldqtys];[typex: [listx: [ident: string]]]];[colonpairlist: [colonpair: [ident: fieldbaseunits];[typex: [listx: [ident: string]]]];[colonpair: [ident: fieldunits];[typex: [listx: [ident: string]]]];[colonpair: [ident: fieldfacts];[typex: [listx: [listx: [ident: string]]]]]]]]
start: 1171; end: 1218 [statestmt: [colonpairlist: [colonpair: [ident: filename];[typex: [ident: string]]];[colonpairlist: [colonpair: [ident: fieldcats];[typex: [listx: [ident: string]]]]]]]
start: 1219; end: 1254 [statestmt: [colonpairlist: [colonpair: [ident: dataTable];[typex: [listx: [listx: [ident: string]]]]]]]
start: 1255; end: 1338 [statestmt: [colonpairlist: [colonpair: [ident: missingCodes];[typex: [listx: [ident: string]]]];[colonpairlist: [colonpair: [ident: anomalous];[typex: [listx: [ident: integer]]]];[colonpair: [ident: excluded];[typex: [listx: [ident: integer]]]]]]]
start: 1340; end: 1502 [respstmt: [ident: fileResponseObj];[responselit: [identlist: [ident: fileResponseIndex];[identlist: [ident: factoids];[ident: missingCodes];[ident: nCols];[ident: dataTable];[ident: fieldnames];[ident: fielddescs];[ident: fieldunits];[ident: fieldcats];[ident: fieldfacts]]]]]
start: 1503; end: 1634 [respstmt: [ident: cleanupResponse];[responselit: [identlist: [ident: fileResponseIndex];[identlist: [ident: factoids];[ident: missingCodes];[ident: nCols];[ident: dataTable];[ident: anomalous];[ident: excludedIndices]]]]]
start: 1676; end: 1756 [transitionstmt: [ident: cleanupExcluded];[strg: "editExcl"];[request: [getpost: [ident: GET]];[identopttyplist: [identopttyp: [ident: linenum];[typex: [ident: integer]]]]];[ident: editExcluded]]
start: 1787; end: 1886 [transitionstmt: [ident: finishCleanup];[strg: "cleanedup"];[request: [getpost: [ident: GET]]];[ident: cleanupDone];[responsex: [respx: [ident: editPage];[ident: fileResponseObj]]]]
start: 1888; end: 1959 [transitionstmt: [ident: CloseFile];[strg: "close"];[request: [getpost: [ident: GET]]];[ident: closeFile];[responsex: [respx: [ident: startPage]]]]
